pkg/log/io/spinner.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2023 The Okteto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io
+
+import (
+	"fmt"
+	"os"
+	"time"
+	"unicode"
+
+	sp "github.com/briandowns/spinner"
+	"golang.org/x/term"
+)
+
+const (
+	// OktetoDisableSpinnerEnvVar if true spinner is disabled
+	OktetoDisableSpinnerEnvVar = "OKTETO_DISABLE_SPINNER"
+
+	// SpinnerAndWhitespaceCharCount is the number of characters that the spinner and the whitespace take
+	SpinnerAndWhitespaceCharCount = 2
+
+	// minimumWidthToTrimSpinnerMsg is the minimum width to trim the spinner message
+	MinimumWidthToTrimSpinnerMsg = 4
+
+	// maxCharsToRemoveForThreeDots is the maximum number of characters to remove to add the three dots
+	CharsToRemoveForThreeDots = 5
+)
+
+// Spinner is the interface for the spinner
+type OktetoSpinner interface {
+	Start()
+	Stop()
+
+	getMessage() string
+}
+
+// ttySpinner is the spinner for the tty
+type ttySpinner struct {
+	*sp.Spinner
+
+	message          string
+	getTerminalWidth func() (int, error)
+}
+
+// newTTYSpinner creates a new ttySpinner
+func newTTYSpinner(message string) *ttySpinner {
+	spinner := sp.New(sp.CharSets[14], 100*time.Millisecond, sp.WithHiddenCursor(true))
+	okSpinner := &ttySpinner{
+		Spinner: spinner,
+		message: ucFirst(message),
+
+		getTerminalWidth: getTerminalWidthFunc,
+	}
+	spinner.PreUpdate = okSpinner.preUpdateFunc()
+	return okSpinner
+}
+
+// Start starts the spinner
+func (s *ttySpinner) Start() {
+	s.Spinner.Start()
+}
+
+// Stop stops the spinner
+func (s *ttySpinner) Stop() {
+	s.Spinner.Stop()
+}
+
+// getMessage returns the spinner message
+func (s *ttySpinner) getMessage() string {
+	return s.message
+}
+
+func (s *ttySpinner) preUpdateFunc() func(spinner *sp.Spinner) {
+	return func(spinner *sp.Spinner) {
+		width, err := s.getTerminalWidth()
+		if err != nil {
+			return
+		}
+		spinner.Suffix = s.calculateSuffix(width)
+	}
+}
+
+func getTerminalWidthFunc() (int, error) {
```

Comment:
Can we test this function too?

Reply:
Nope, `os.Stdout.Fd` is only working when you have attached it to a terminal, so the only thing we could test is that the value when there is no stdout is 0 which is an scenario that will never happen, so testing it won't prevent any bug and it can only cause that when a bug occur on that function we can think that the test is already contemplating it and could cause some missunderstanding